core/templates/dev/head/pages/exploration_editor/history_tab/HistoryServices.js

```diff
@@ -198,12 +198,23 @@ oppia.factory('compareVersionsService', [
           v1Data: $http.get(explorationDataUrl + v1),
           v2Data: $http.get(explorationDataUrl + v2)
         }).then(function(response) {
-          var v1States = response.v1Data.data.states;
-          var v2States = response.v2Data.data.states;
+          var v1StatesDict = response.v1Data.data.states;
+          var v2StatesDict = response.v2Data.data.states;
 
           // Track changes from v1 to LCA, and then from LCA to v2.
           var lca = versionsTreeService.findLCA(v1, v2);
 
+          var v1States = {};
+          var v2States = {};
+          for (var stateName in v1StatesDict) {
```

Comment:
Do you mean just putting a function into StateObjectFactory for taking in a list of state dicts and returning the state objects?